pytorch_lightning/trainer/trainer.py

```diff
@@ -1396,3 +1397,11 @@ def _on_exception(self):
         # save a checkpoint for fault tolerant training. we don't use `log_dir` to minimize the chances of failure.
         file_path = os.path.join(self.default_root_dir, ".pl_auto_save.ckpt")
         self.save_checkpoint(file_path)
+
+    @contextmanager
+    def _evaluation_context(self) -> Generator:
+        with torch.inference_mode() if _TORCH_GREATER_EQUAL_1_9 and self.state.fn in (
+            TrainerFn.TESTING,
+            TrainerFn.PREDICTING,
+        ) else torch.no_grad():
```

Comment:
```suggestion
        with torch.inference_mode() if _TORCH_GREATER_EQUAL_1_9 else torch.no_grad():
```
I don't see a reason to limit this to test and predict only. 
Should we move this context manager out of the trainer? It does not depend on the state, only the pytorch version.

Reply:
> I don't see a reason to limit this to test and predict only.

+1, see my previous comment: https://github.com/PyTorchLightning/pytorch-lightning/pull/8813#pullrequestreview-730040781

Reply:
+1 for your previous comment, I did not see it

Reply:
> Could we get the benchmark for some traditional models to validate this doesn't bring a regression for Lightning performances + updated docs.

Would you mind proposing some traditional models that you have in mind? I've done [some benchmark tests](https://colab.research.google.com/drive/1GysINpTTt9g55r3XG4z3vdW458sopO7W?usp=sharing), although they don't follow previous examples as I couldn't find any good one.

Reply:
@carmocca  @awaelchli Okay, we can remove the `self.state.fn` check. @ananthsub and I were just a bit more cautious about turning on the inference mode during validation.

Reply:
@tangbinh keep in mind the `trainer.fit` validation and the `trainer.validate` validation used different code paths.

In this PR you are just applyint the nee context manager to the latter

Reply:
```suggestion
        with torch.inference_mode() if _TORCH_GREATER_EQUAL_1_9 else torch.no_grad():
```